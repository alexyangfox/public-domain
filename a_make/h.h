/*************************************************************************
|                                                                        |
|   H.H                                                         30.09.89 |
|   PAMAKE Utility:  header file                                         |
|                                                                        |
*************************************************************************/

#define uchar       unsigned char
#define bool        uchar
#define TRUE        (1)
#define FALSE       (0)
#define DEFN1       "makefile"              /*  Default names  */
#define LIFNAME     "TEMP_LIF.TMP"
#ifdef  VMS
#define LZ          (4096)                  /*  Line size  */
#else
#define LZ          (1024)                  /*  Line size  */
#endif
#define DOLLAR      1                       /* special dollar processing */

#ifdef VMS
#define PXNORMAL 1          /* normal exit() arg */
#define PXERROR  0          /* error exit() arg */
#define unlink(filename) remove(filename)
#else
#define PXNORMAL 0          /* normal exit() arg */
#define PXERROR  1          /* error exit() arg */
#define PFPROTO  1          /* use function prototypes */
#endif

#ifdef __TSC__
#define PAMDOS 1
#endif

#ifdef __TURBOC__
#define PAMAKE_FTIME 1
#define PAMDOS 1
#endif

#ifdef LATTICE
#define PAMAKE_FTIME 1
#define PAMOS2 1
#endif

#ifndef DOLLAR
#define dollar(p) (p)       /* dont bother to strip special dollars */
#endif

struct name
{
    struct name *           n_next;         /* Next in the list of names */
    char *                  n_name;         /* Called */
    struct line *           n_line;         /* Dependencies */
    long                    n_time;         /* Modify time of this name */
    uchar                   n_flag;         /* Info about the name */
};

#define N_MARK              0x01            /* For cycle check */
#define N_DONE              0x02            /* Name looked at */
#define N_TARG              0x04            /* Name is a target */
#define N_PREC              0x08            /* Target is precious */
#define N_DYND              0x10            /* Name was made by dyndep */
#define N_DOUBLE            0x20            /* Double colon */

struct  line
{
    struct line *           l_next;         /* Next line (for ::) */
    struct depend *         l_dep;          /* Dependents for this line */
    struct cmd *            l_cmd;          /* Commands for this line */
};

struct  depend
{
    struct depend *         d_next;         /* Next dependent */
    struct name *           d_name;         /* Name of dependent */
};

struct  cmd
{
    struct cmd *            c_next;         /* Next command line */
    struct lif *            c_lif;          /* LIF lines for this command */
    char *                  c_cmd;          /* Command line */
};

struct  lif
{
    struct lif *            f_next;         /* Next LIF line */
    char *                  f_lif;          /* LIF line */
};

struct  macro
{
    struct macro *          m_next;         /* Next variable */
    char *                  m_name;         /* Called ... */
    char *                  m_val;          /* Its value */
    char *                  m_sub;          /* Temp subst value */
    uchar                   m_flag;         /* Infinite loop check */
};

#define M_LOOPCHK           0x01            /* For loop check */
#define M_ENV               0x02            /* ex-environment macro */
#define M_PERM              0x04            /* not resetable */

extern char *           myname;
extern struct name      namehead;
extern struct macro *   macrohead;
extern struct name *    firstname;
extern bool             silent;
extern bool             confirm;
extern bool             ignore;
extern bool             rules;
extern bool             dotouch;
extern bool             quest;
extern bool             domake;
extern bool             display;
extern char             str1[];
extern char             str2[];
extern int              lineno;
extern uchar            macrotype;
extern unsigned char    pamakeos2;

extern FILE *           ifile[4];
extern int              fln[4];
extern char             fname[4][80];
extern int              nestlvl;

#ifdef PFPROTO
extern  void            check(struct name *np);
extern  void            checklif(struct cmd *cp);
extern  void            circh(void);
extern  void            cleardynflag(struct name *np);
extern  unsigned long   curtime(void);
extern  void            docmds1(struct name *np,struct line *lp);
extern  void            docmds(struct name *np);
extern  void            dodisp(char *name,long t);
extern  void            doexp(char * *to,char *from,int *len,char *buf);
extern  char *          dollar(char *);
extern  void            dosetcmd(char *p);
extern  void            dostatcmd(char *p);
extern  int             dosh(char *p,int shell);
extern  int             dos_internal(char *s);
extern  bool            dyndep(struct name *np);
extern  void            expand(char *str);
extern  void            ifeoc(void);
extern  void            ifeof(void);
extern  int             ifproc(char *s,int n);
extern  void            input(void);
extern  int             istrue(char *s);
extern  bool            getline(char *str);
extern  struct macro *  getmp(char *name);
extern  char *          gettok(char * *ptr);
extern  void            killlif(void);
extern  void            main(int argc,char * *argv,char * *envp);
extern  int             make(struct name *np,int level);
extern  void            make1(struct name *np,struct line *lp,struct depend *qdp);
extern  void            makelif(struct cmd *cp);
extern  void            makerules(void);
extern  void            markmacros(void);
extern  void            modtime(struct name *np);
extern  struct name *   newname(char *name);
extern  struct depend * newdep(struct name *np,struct depend *dp);
extern  struct cmd *    newcmd(char *str,struct cmd *cp,struct cmd * *crp);
extern  struct lif *    newlif(char *str,struct lif *lp);
extern  void            newline(struct name *np,struct depend *dp,struct cmd *cp,int flag);
extern  void            precious(void);
extern  void            prt(void);
extern  char *          pstrstr(char *, char *);
extern  void            setdmacros(struct name *np,struct depend *qdp);
extern  struct macro *  setmacro(char *name,char *val);
extern  char *          suffix(char *name);
extern  void            touch(struct name *np);
extern  void            usage(void);
extern  int             pspace(int);
#ifdef __TSC__
#pragma save, call (reg_param => ())
extern  void            error(char *msg,...);
extern  void            fatal(char *msg,...);
#pragma restore
#else
extern  void            error(char *msg,...);
extern  void            fatal(char *msg,...);
#endif
#else
unsigned long           curtime();
char *                  dollar();
void                    error();
void                    fatal();
char *                  getmacro();
struct macro *          getmp();
char *                  gettok();
void                    input();
int                     make();
void                    makerules();
struct cmd *            newcmd();
struct depend *         newdep();
void                    newline();
struct name *           newname();
void                    precious();
char *                  pstrstr();
struct macro *          setmacro();
char *                  suffix();
void                    touch();
#endif
