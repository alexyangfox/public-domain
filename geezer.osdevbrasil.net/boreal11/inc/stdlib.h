#ifndef __TL_STDLIB_H
#define	__TL_STDLIB_H

#ifdef __cplusplus
extern "C"
{
#endif

unsigned rand(void);
void srand(unsigned new_seed);
void exit(int exit_code);

#ifdef __cplusplus
}
#endif

#endif
