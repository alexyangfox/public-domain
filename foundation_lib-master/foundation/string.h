/* utf.h  -  Foundation library  -  Public Domain  -  2013 Mattias Jansson / Rampant Pixels
 * 
 * This library provides a cross-platform foundation library in C11 providing basic support data types and
 * functions to write applications and games in a platform-independent fashion. The latest source code is
 * always available at
 * 
 * https://github.com/rampantpixels/foundation_lib
 * 
 * This library is put in the public domain; you can redistribute it and/or modify it without any restrictions.
 *
 */

#pragma once

/*! \file utf.h
    String handling and utility functions in UTF-8, conversion to/from UTF-16 */

#include <foundation/platform.h>
#include <foundation/types.h>


FOUNDATION_API char*          string_allocate( unsigned int length );
FOUNDATION_API void           string_deallocate( char* str );
FOUNDATION_API char*          string_clone( const char* str );
FOUNDATION_API char*          string_format( const char* format, ... );
FOUNDATION_API char*          string_format_buffer( char* buffer, unsigned int maxlen, const char* format, ... );
FOUNDATION_API char*          string_vformat( const char* format, va_list list );
FOUNDATION_API unsigned int   string_length( const char* str );
FOUNDATION_API unsigned int   string_glyphs( const char* str );
FOUNDATION_API hash_t         string_hash( const char* str );

FOUNDATION_API char*          string_resize( char* str, unsigned int length, char c );

//! Like strncpy, except that dst will always be zero terminated (i.e copies at most limit-1 characters from src)
FOUNDATION_API void           string_copy( char* dst, const char* src, unsigned int limit );

//! These methods modify the passed string and returns a modified string pointer
FOUNDATION_API char*          string_strip( char* str, const char* delimiters );
FOUNDATION_API char*          string_strip_substr( char* str, const char* delimiters, unsigned int length );
FOUNDATION_API char*          string_replace( char* str, const char* key, const char* newkey, bool repeat );
FOUNDATION_API char*          string_append( char* str, const char* suffix );
FOUNDATION_API char*          string_prepend( char* str, const char* prefix );

//! These methods do not modify the passed string and returns a new string pointer
FOUNDATION_API char*          string_concat( const char* lhs, const char* rhs );
FOUNDATION_API void           string_split( const char* str, const char* separators, char** left, char** right, bool allowempty );
FOUNDATION_API char*          string_substr( const char* str, unsigned int offset, unsigned int length );

FOUNDATION_API unsigned int   string_find( const char* str, char c, unsigned int offset );
FOUNDATION_API unsigned int   string_find_string( const char* str, const char* key, unsigned int offset );
FOUNDATION_API unsigned int   string_rfind( const char* str, char c, unsigned int offset );
FOUNDATION_API unsigned int   string_rfind_string( const char* str, const char* key, unsigned int offset );
FOUNDATION_API unsigned int   string_find_first_of( const char* str, const char* tokens, unsigned int offset );
FOUNDATION_API unsigned int   string_find_last_of( const char* str, const char* tokens, unsigned int offset );
FOUNDATION_API unsigned int   string_find_first_not_of( const char* str, const char* tokens, unsigned int offset );
FOUNDATION_API unsigned int   string_find_last_not_of( const char* str, const char* tokens, unsigned int offset );
FOUNDATION_API unsigned int   string_rfind_first_of( const char* str, const char* tokens, unsigned int offset );
FOUNDATION_API unsigned int   string_rfind_first_not_of( const char* str, const char* tokens, unsigned int offset );

FOUNDATION_API bool           string_ends_with( const char* str, const char* suffix );

FOUNDATION_API bool           string_equal( const char* lhs, const char* rhs );
FOUNDATION_API bool           string_equal_substr( const char* lhs, const char* rhs, unsigned int length );

FOUNDATION_API bool           string_match_pattern( const char* element, const char* pattern );

FOUNDATION_API char**         string_explode( const char* str, const char* delimiters, bool allow_empty );
FOUNDATION_API char*          string_merge( const char* const* array, unsigned int array_size, const char* delimiter );

FOUNDATION_API uint32_t       string_glyph( const char* str, unsigned int offset, unsigned int* consumed );

FOUNDATION_API int            string_array_find( const char* const* haystack, const char* needle, unsigned int haystack_size );

FOUNDATION_API void           string_array_deallocate_elements( char** array );
#define                       string_array_deallocate( array ) ( string_array_deallocate_elements(array), array_deallocate(array), array=0 )

FOUNDATION_API wchar_t*       wstring_allocate_from_string( const char* cstr, unsigned int length );
FOUNDATION_API void           wstring_from_string( wchar_t* str, const char* cstr, unsigned int max_length );
FOUNDATION_API void           wstring_deallocate( wchar_t* str );
FOUNDATION_API unsigned int   wstring_length( const wchar_t* str );
FOUNDATION_API bool           wstring_equal( const wchar_t* lhs, const wchar_t* rhs );

FOUNDATION_API char*          string_allocate_from_wstring( const wchar_t* str, unsigned int length );
FOUNDATION_API char*          string_allocate_from_utf16( const uint16_t* str, unsigned int length );
FOUNDATION_API char*          string_allocate_from_utf32( const uint32_t* str, unsigned int length );

FOUNDATION_API void           string_convert_utf16( char* dst, const uint16_t* src, unsigned int dstsize, unsigned int srclength );
FOUNDATION_API void           string_convert_utf32( char* dst, const uint32_t* src, unsigned int dstsize, unsigned int srclength );


/*! Convert an integer to a string, with optional hexadecimal base and base prefix, field width and fill character
    \param val                Integer value
    \param width              Field width
    \param padding            Fill character */
FOUNDATION_API char*          string_from_int( int64_t val, unsigned int width, char padding );

/*! Convert an unsigned integer to a string, with optional hexadecimal base and base prefix, field width and fill character
    \param val                Integer value
    \param hex                Hexadecimal flag
    \param width              Field width
    \param padding            Fill character */
FOUNDATION_API char*          string_from_uint( uint64_t val, bool hex, unsigned int width, char padding );

/*! Convert an 128-bit unsigned integer to a string
    \param val                Integer value */
FOUNDATION_API char*          string_from_uint128( const uint128_t val );

/*! Convert a float to a string, with optional fixed notation, field width, precision and fill character
    \param val                Float value
    \param precision          Precision
    \param width              Field width
    \param padding            Fill character */
FOUNDATION_API char*          string_from_real( real val, unsigned int precision, unsigned int width, char padding );

/*! Get a formatted string of the given timestamp
    \param time               Timestamp
    \return                   String */
FOUNDATION_API char*          string_from_time( uint64_t time );

/*! Get a formatted string of the given UUID
    \param uuid               UUID
    \return                   String */
FOUNDATION_API char*          string_from_uuid( const uuid_t uuid );

/*! Get a formatted string of the given version
    \param version            Version
    \return                   String */
FOUNDATION_API char*          string_from_version( const version_t version );

FOUNDATION_API char*          string_from_int_buffer( char* buffer, int64_t val, unsigned int width, char padding );
FOUNDATION_API char*          string_from_uint_buffer( char* buffer, uint64_t val, bool hex, unsigned int width, char padding );
FOUNDATION_API char*          string_from_uint128_buffer( char* buffer, const uint128_t val );
FOUNDATION_API char*          string_from_real_buffer( char* buffer, real val, unsigned int precision, unsigned int width, char padding );
FOUNDATION_API char*          string_from_time_buffer( char* buffer, uint64_t time );
FOUNDATION_API char*          string_from_uuid_buffer( char* buffer, const uuid_t uuid );
FOUNDATION_API char*          string_from_version_buffer( char* buffer, const version_t version );

FOUNDATION_API const char*    string_from_int_static( int64_t val, unsigned int width, char padding );
FOUNDATION_API const char*    string_from_uint_static( uint64_t val, bool hex, unsigned int width, char padding );
FOUNDATION_API const char*    string_from_uint128_static( const uint128_t val );
FOUNDATION_API const char*    string_from_real_static( real val, unsigned int precision, unsigned int width, char padding );
FOUNDATION_API const char*    string_from_time_static( uint64_t time );
FOUNDATION_API const char*    string_from_uuid_static( const uuid_t uuid );
FOUNDATION_API const char*    string_from_version_static( const version_t version );


/*! Convert a string to a signed integer
    \param val                String */
FOUNDATION_API int            string_to_int( const char* val );

/*! Convert a string to an unsigned integer
    \param val                String
    \param hex                Hexadecimal flag */
FOUNDATION_API unsigned int   string_to_uint( const char* val, bool hex );

/*! Convert a string to a signed 64-bit integer
    \param val                String */
FOUNDATION_API int64_t        string_to_int64( const char* val );

/*! Convert a string to an unsigned 64-bit integer
    \param val                String
    \param hex                Hexadecimal flag */
FOUNDATION_API uint64_t       string_to_uint64( const char* val, bool hex );

/*! Convert a string to an unsigned 128-bit integer
    \param val                String */
FOUNDATION_API uint128_t      string_to_uint128( const char* val );

/*! Convert a string to a float
    \param val                String */
FOUNDATION_API real           string_to_real( const char* val );

/*! Convert a string to a uuid
    \param val                String */
FOUNDATION_API uuid_t         string_to_uuid( const char* str );

/*! Convert a string to a version identifier
    \param val                String */
FOUNDATION_API version_t      string_to_version( const char* str );

/*! Validate email address format (basic tests)
    \param address            Email address string */
FOUNDATION_API bool           string_is_valid_email_address( const char* address );


#define STRING_NPOS                     0xFFFFFFFFU

#define STRING_WHITESPACE               " \n\r\t"
#define WSTRING_WHITESPACE             L" \n\r\t"

#if FOUNDATION_PLATFORM_REALSIZE == 64
#  define STRING_FORMAT_REAL            "llf"
#else
#  define STRING_FORMAT_REAL            "f"
#endif

#if FOUNDATION_PLATFORM_WINDOWS
#  if FOUNDATION_PLATFORM_POINTER_SIZE == 8
#    define STRING_FORMAT_POINTER       "0x%016p"
#  else
#    define STRING_FORMAT_POINTER       "0x%08p"
#  endif
#else
#  if FOUNDATION_PLATFORM_POINTER_SIZE == 8
#    define STRING_FORMAT_POINTER       "%016p"
#  else
#    define STRING_FORMAT_POINTER       "%08p"
#  endif
#endif
