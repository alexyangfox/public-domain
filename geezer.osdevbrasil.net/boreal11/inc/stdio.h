#ifndef __TL_STDIO_H
#define	__TL_STDIO_H

#ifdef __cplusplus
extern "C"
{
#endif

#include <_va_list.h>
#include <_null.h>
#include <_gcc.h> /* __PRINTF0__, __PRINTF1__ */

#define	_IONBF	0	/* unbuffered stream */
#define _IOFBF  1	/* fully buffered stream */
#define _IOLBF  2	/* line and fully buffered stream */

#define	stdin	(g_std_streams + 0)
#define	stdout	(g_std_streams + 1)
//#define	stderr	(g_std_streams + 2)

#define	EOF	(-1)

/* size of default stdio.h buffers -- a nice, random
value which I hope will expose any bugs in the libc code */
#define BUFSIZ	53

#define	putchar(char)	fputc(char, stdout)

typedef struct
{
	char *buf_base, *buf_ptr;
	unsigned size, room, flags;
	int handle;
} FILE;

extern FILE g_std_streams[];

int fflush(FILE *stream);
int fputc(int c, FILE *stream);
int fputs(const char *str, FILE *stream);
int vsprintf(char *buffer, const char *fmt, va_list args);
void setbuf(FILE *stream, char *buffer);
int vprintf(const char *fmt, va_list args);

int sprintf(char *buffer, const char *fmt, ...) __PRINTF1__;
int printf(const char *fmt, ...) __PRINTF0__;

#ifdef __cplusplus
}
#endif

#endif
