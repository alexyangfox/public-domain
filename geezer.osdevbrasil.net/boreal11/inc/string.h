#ifndef __TL_STRING_H
#define	__TL_STRING_H

#ifdef __cplusplus
extern "C"
{
#endif

#include <_size_t.h>
#include <_null.h>

int memcmp(const void *left_p, const void *right_p, size_t count);
void *memcpy(void *dst_ptr, const void *src_ptr, size_t count);
void *memsetw(void *dst, int val, size_t count);
void *memset(void *dst, int val, size_t count);

size_t strlen(const char *str);
int strcmp(const char *s1, const char*s2);
int strncmp(const char *s1, const char*s2, unsigned max_len);

#ifdef __cplusplus
}
#endif

#endif

