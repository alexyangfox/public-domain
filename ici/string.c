#define ICI_CORE
#include "fwd.h"
#include "str.h"
#include "struct.h"
#include "exec.h"
#include "int.h"
#include "primes.h"

/*
 * How many bytes of memory we need for a string of n chars (single
 * allocation).
 */
#define STR_ALLOCZ(n)   (offsetof(ici_str_t, s_u) + (n) + 1)

/*
 * Allocate a new string object (single allocation) large enough to hold
 * nchars characters, and register it with the garbage collector.  Note: This
 * string is not yet an atom, but must become so as it is *not* mutable.
 *
 * WARINING: This is *not* the normal way to make a string object. See
 * ici_str_new().
 *
 * This --func-- forms part of the --ici-api--.
 */
ici_str_t *
ici_str_alloc(int nchars)
{
    register ici_str_t  *s;
    size_t              az;

    az = STR_ALLOCZ(nchars);
    if ((s = (ici_str_t *)ici_nalloc(az)) == NULL)
        return NULL;
    ICI_OBJ_SET_TFNZ(s, TC_STRING, 0, 1, az <= 127 ? az : 0);
    s->s_chars = s->s_u.su_inline_chars;
    s->s_nchars = nchars;
    s->s_chars[nchars] = '\0';
    s->s_struct = NULL;
    s->s_slot = NULL;
#   if KEEP_STRING_HASH
        s->s_hash = 0;
#   endif
    s->s_vsver = 0;
    ici_rego(s);
    return s;
}

/*
 * Make a new atomic immutable string from the given characters.
 *
 * Note that the memory allocated to a string is always at least one byte
 * larger than the listed size and the extra byte contains a '\0'.  For
 * when a C string is needed.
 *
 * The returned string has a reference count of 1 (which is caller is
 * expected to decrement, eventually).
 *
 * See also: 'ici_str_new_nul_term()' and 'ici_str_get_nul_term()'.
 *
 * Returns NULL on error, usual conventions.
 *
 * This --func-- forms part of the --ici-api--.
 */
ici_str_t *
ici_str_new(char *p, int nchars)
{
    ici_str_t           *s;
    size_t              az;
    static struct
    {
        ici_str_t       s;
        char            d[40];
    }
        proto   = {{OBJ(TC_STRING)}};

    assert(nchars >= 0);
    az = STR_ALLOCZ(nchars);
    if ((size_t)nchars < sizeof proto.d)
    {
        ici_obj_t       **po;

        proto.s.s_nchars = nchars;
        proto.s.s_chars = proto.s.s_u.su_inline_chars;
        memcpy(proto.s.s_chars, p, nchars);
        proto.s.s_chars[nchars] = '\0';
#       if KEEP_STRING_HASH
            proto.s.s_hash = 0;
#       endif
        if ((s = stringof(atom_probe(objof(&proto.s), &po))) != NULL)
        {
            ici_incref(s);
            return s;
        }
        ++ici_supress_collect;
        az = STR_ALLOCZ(nchars);
        if ((s = (ici_str_t *)ici_nalloc(az)) == NULL)
            return NULL;
        memcpy((char *)s, (char *)&proto.s, az);
        ICI_OBJ_SET_TFNZ(s, TC_STRING, O_ATOM, 1, az);
        s->s_chars = s->s_u.su_inline_chars;
        ici_rego(s);
        --ici_supress_collect;
        ICI_STORE_ATOM_AND_COUNT(po, s);
        return s;
    }
    if ((s = (ici_str_t *)ici_nalloc(az)) == NULL)
        return NULL;
    ICI_OBJ_SET_TFNZ(s, TC_STRING, 0, 1, az <= 127 ? az : 0);
    s->s_chars = s->s_u.su_inline_chars;
    s->s_nchars = nchars;
    s->s_struct = NULL;
    s->s_slot = NULL;
    s->s_vsver = 0;
    memcpy(s->s_chars, p, nchars);
    s->s_chars[nchars] = '\0';
#   if KEEP_STRING_HASH
        s->s_hash = 0;
#   endif
    ici_rego(s);
    return stringof(ici_atom(objof(s), 1));
}

/*
 * Make a new atomic immutable string from the given nul terminated
 * string of characters.
 *
 * The returned string has a reference count of 1 (which is caller is
 * expected to decrement, eventually).
 *
 * Returns NULL on error, usual conventions.
 *
 * This --func-- forms part of the --ici-api--.
 */
ici_str_t *
ici_str_new_nul_term(char *p)
{
    register ici_str_t  *s;

    if ((s = ici_str_new(p, strlen(p))) == NULL)
        return NULL;
    return s;
}

/*
 * Make a new atomic immutable string from the given nul terminated
 * string of characters.
 *
 * The returned string has a reference count of 0, unlike
 * ici_str_new_nul_term() which is exactly the same in other respects.
 *
 * Returns NULL on error, usual conventions.
 *
 * This --func-- forms part of the --ici-api--.
 */
ici_str_t *
ici_str_get_nul_term(char *p)
{
    ici_str_t   *s;

    if ((s = ici_str_new(p, strlen(p))) == NULL)
        return NULL;
    ici_decref(s);
    return s;
}

/*
 * Return a new mutable string (i.e. one with a seperate growable allocation).
 * The initially allocated space is n, but the length is 0 until it has been
 * set by the caller.
 *
 * The returned string has a reference count of 1 (which is caller is
 * expected to decrement, eventually).
 *
 * Returns NULL on error, usual conventions.
 *
 * This --func-- forms part of the --ici-api--.
 */
ici_str_t *
ici_str_buf_new(int n)
{
    ici_str_t           *s;

    if ((s = ici_talloc(ici_str_t)) == NULL)
        return NULL;
    if ((s->s_chars = ici_nalloc(n)) == NULL)
    {
        ici_tfree(s, ici_str_t);
        return NULL;
    }
    ICI_OBJ_SET_TFNZ(s, TC_STRING, ICI_S_SEP_ALLOC, 1, 0);
    s->s_u.su_nalloc = n;
    s->s_vsver = 0;
    s->s_nchars = 0;
    s->s_struct = NULL;
    s->s_slot = NULL;
    s->s_vsver = 0;
    return s;
}

/*
 * Ensure that the given string has enough allocated memory to hold a string
 * of n characters (and a guard '\0' which this routine stores).  Grows ths
 * string as necessary.  Returns 0 on success, 1 on error, usual conventions.
 * Checks that the string is mutable and not atomic.
 *
 * This --func-- forms part of the --ici-api--.
 */
int
ici_str_need_size(ici_str_t *s, int n)
{
    char                *chars;
    char                n1[30];

    if ((s->o_head.o_flags & (O_ATOM|ICI_S_SEP_ALLOC)) != ICI_S_SEP_ALLOC)
    {
        sprintf(ici_buf, "attempt to modify an atomic string %s", ici_objname(n1, objof(s)));
        ici_error = ici_buf;
        return 1;
    }
    if (s->s_u.su_nalloc >= n + 1)
        return 0;
    n <<= 1;
    if ((chars = ici_nalloc(n)) == NULL)
        return 1;
    memcpy(chars, s->s_chars, s->s_nchars + 1);
    ici_nfree(s->s_chars, s->s_u.su_nalloc);
    s->s_chars = chars;
    s->s_u.su_nalloc = n;
    s->s_chars[n >> 1] = '\0';
    return 0;
}

/*
 * Mark this and referenced unmarked objects, return memory costs.
 * See comments on t_mark() in object.h.
 */
static unsigned long
mark_string(ici_obj_t *o)
{
    o->o_flags |= O_MARK;
    return STR_ALLOCZ(stringof(o)->s_nchars);
}

/*
 * Returns 0 if these objects are equal, else non-zero.
 * See the comments on t_cmp() in object.h.
 */
static int
cmp_string(ici_obj_t *o1, ici_obj_t *o2)
{
    if (stringof(o1)->s_nchars != stringof(o2)->s_nchars)
        return 1;
    return memcmp
    (
        stringof(o1)->s_chars,
        stringof(o2)->s_chars,
        stringof(o1)->s_nchars
    );
}

/*
 * Return a copy of the given object, or NULL on error.
 * See the comment on t_copy() in object.h.
 */
static ici_obj_t *
copy_string(ici_obj_t *o)
{
    ici_str_t           *ns;

    if ((ns = ici_str_buf_new(stringof(o)->s_nchars + 1)) == NULL)
        return NULL;
    ns->s_nchars = stringof(o)->s_nchars;
    memcpy(ns->s_chars, stringof(o)->s_chars, ns->s_nchars);
    ns->s_chars[ns->s_nchars] = '\0';
    return objof(ns);
}

/*
 * Free this object and associated memory (but not other objects).
 * See the comments on t_free() in object.h.
 */
static void
free_string(ici_obj_t *o)
{
    if (o->o_flags & ICI_S_SEP_ALLOC)
        ici_nfree(stringof(o)->s_chars, stringof(o)->s_u.su_nalloc);
    ici_nfree(o, STR_ALLOCZ(stringof(o)->s_nchars));
}

/*
 * Return a hash sensitive to the value of the object.
 * See the comment on t_hash() in object.h
 */
unsigned long
ici_hash_string(ici_obj_t *o)
{
    unsigned long       h;

#   if KEEP_STRING_HASH
        if (stringof(o)->s_hash != 0)
            return stringof(o)->s_hash;
#   endif
    h = ici_crc(STR_PRIME_0, (const unsigned char *)stringof(o)->s_chars, stringof(o)->s_nchars);
#   if KEEP_STRING_HASH
        stringof(o)->s_hash = h;
#   endif
    return h;
}

/*
 * Return the object at key k of the obejct o, or NULL on error.
 * See the comment on t_fetch in object.h.
 */
static ici_obj_t *
fetch_string(ici_obj_t *o, ici_obj_t *k)
{
    register int        i;

    if (!isint(k))
        return ici_fetch_fail(o, k);
    if ((i = (int)intof(k)->i_value) < 0 || i >= stringof(o)->s_nchars)
        k = objof(ici_str_new("", 0));
    else
        k = objof(ici_str_new(&stringof(o)->s_chars[i], 1));
    if (k != NULL)
        ici_decref(k);
    return k;
}

/*
 * Assign to key k of the object o the value v. Return 1 on error, else 0.
 * See the comment on t_assign() in object.h.
 *
 * The key k must be a positive integer. The string will attempt to grow
 * to accomodate the new index as necessary.
 */
static int
assign_string(ici_obj_t *o, ici_obj_t *k, ici_obj_t *v)
{
    long        i;
    long        n;
    ici_str_t   *s;

    if (o->o_flags & O_ATOM)
    {
        ici_error = "attempt to assign to an atomic string";
        return 1;
    }
    if (!isint(k) || !isint(v))
        return ici_assign_fail(o, k, v);
    i = intof(k)->i_value;
    if (i < 0)
    {
        ici_error = "attempt to assign to negative string index";
        return 1;
    }
    s = stringof(o);
    if (ici_str_need_size(s, i + 1))
        return 1;
    for (n = s->s_nchars; n < i; ++n)
        s->s_chars[n] = ' ';
    s->s_chars[i] = (char)intof(v)->i_value;
    if (s->s_nchars < ++i)
    {
        s->s_nchars = i;
        s->s_chars[i] = '\0';
    }
    return 0;
}


ici_type_t  string_type =
{
    mark_string,
    free_string,
    ici_hash_string,
    cmp_string,
    copy_string,
    assign_string,
    fetch_string,
    "string"
};
