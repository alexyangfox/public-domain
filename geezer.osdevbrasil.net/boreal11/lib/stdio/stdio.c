#include <stdio.h> /* BUFSIZ, FILE, NULL, _IOLBF */

static char g_stdout_buffer[BUFSIZ];

FILE g_std_streams[] =
{
	{	/* buf_base	buf_ptr		size	room	flags	handle */
/* stdin */	NULL,		NULL,		0,	0,	0,	0
	},
	{
/* stdout */	g_stdout_buffer, g_stdout_buffer, BUFSIZ, BUFSIZ, _IOLBF, 1
	},
	{
/* stderr */	NULL,		NULL,		0,	0,	0,	2
	}
};
