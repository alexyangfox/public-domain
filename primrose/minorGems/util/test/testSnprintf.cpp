/*
 * Modification History
 *
 * 2004-January-15   Jason Rohrer
 * Created.
 */

/**
 * A test program for snprintf behavior.
 */


#include <stdio.h>
#include <stdlib.h>
#include <string.h>



int main() { 
    int numStrings = 3;
    
    // test strings of length 3, 4, and 5
    const char *testStrings[3] = { "tst", "test", "testt" };

    int result;
    
    // a buffer of length 4, which IS NOT large
    // enough to hold the last two testStrings
    char *buffer = (char*)( malloc( 4 ) );

    int i;
    
    for( i=0; i<numStrings; i++ ) {
        // clear buffer with 'a' characters
        memset( (void*)buffer, (int)( 'a' ), 4 );
    
        // print testStringA into buffer
        result = snprintf( buffer, 4, "%s", testStrings[i] );

        printf( "Printed string of length %d to buffer of "
                "size 4 with snprintf.\n"
                "Return value = %d\n",
                strlen( testStrings[i] ),
                result );

        if( buffer[3] == '\0' ) {
            printf( "Buffer was null-terminated by snprintf\n\n" );
            }
        else {
            printf( "Buffer was NOT null-terminated by snprintf\n\n" );
            }
        }

    free( buffer );
    
    return 0;
}

