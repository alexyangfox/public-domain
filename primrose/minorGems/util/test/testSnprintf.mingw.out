Printed string of length 3 to buffer of size 4 with snprintf.
Return value = 3
Buffer was null-terminated by snprintf

Printed string of length 4 to buffer of size 4 with snprintf.
Return value = 4
Buffer was NOT null-terminated by snprintf

Printed string of length 5 to buffer of size 4 with snprintf.
Return value = -1
Buffer was NOT null-terminated by snprintf

